### DISCUSSION TITLE
Understanding the Contextual Scope of GitHub Copilot

### DISCUSSION THREAD

**Discussion by sakashita-mk** [community/community#69280] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

Hello,

I've been exploring GitHub Copilot and have been impressed with its autocomplete-style suggestions. According to GitHub's official documentation:

> "GitHub Copilot is an AI pair programmer that offers autocomplete-style suggestions as you code. You can receive suggestions from GitHub Copilot either by starting to write the code you want to use, or by writing a natural language comment describing what you want the code to do. GitHub Copilot analyzes the context in the file you are editing, as well as related files, and offers suggestions from within your text editor. GitHub Copilot is powered by a generative AI model developed by GitHub, OpenAI, and Microsoft."
https://docs.github.com/en/copilot/overview-of-github-copilot/about-github-copilot-for-individuals

While this is quite informative, I'm intrigued by what is meant by "the context in the file you are editing, as well as related files." Specifically:

- Does the context include all files that are currently open in tabs?
- If a file imports a custom package, does Copilot consider the content of the imported file as context?
- Are all files in the same directory as the file being edited considered related?

I understand that the real behavior might depend on several factors, but I'd appreciate any insights or hypotheses you all might have on this. Understanding the depth and breadth of Copilot's context can significantly influence how we utilize it in our coding workflows.

Thank you for sharing your experiences and thoughts!

**Comment by Symbolk** [comment]
Maybe it is a vital business secret, hh

**Comment by ayim** [comment, marked as answer]
https://github.blog/2023-05-17-how-github-copilot-is-getting-better-at-understanding-your-code/

TLDR: before your cursor, after your cursor, and in some scenarios, other open files.

**Comment by mrosata** [comment]
From my understanding, Copilot will use the current file and any open tabs in the IDE as the context to generate autocomplete suggestions.  Today I realized that this scope includes other projects also open in the IDE _(at least in JetBrains WebStorm this is true)_. So if I have 2 projects open in my IDE, it seems that I get suggestions in project 1 based on code opened in project 2.

I'm curious if reading the other projects code is the intended copilot behavior or if it is WebStorm leaking context to copilot.

_It would be nice to be able to configure projects to not share code from opened tabs with another project that is also using copilot or to isolate/group projects/workspaces._

**Reply by ayim** [reply]
@Akash1134

**Reply by Akash1134** [reply]
Hi @mrosata ,

Copilot is intended to generate suggestions based on the context of the code you're writing, which includes the currently active file in your editor. It is not designed to explicitly consider code from other projects that are currently open in your IDE. The behaviour you're experiencing could be related to how WebStorm manages and presents project contexts to extensions, which might cause Copilot to consider code from other open projects.

Thanks!

**Comment by pashamed** [comment]
https://learn.microsoft.com/en-us/visualstudio/ide/copilot-chat-context?view=vs-2022

Some improvements of context awareness

**Reply by ayim** [reply]
Yes! We're doing a lot of thinking in the context space for GitHub Copilot Chat, stay tuned.
